### DISCUSSION TITLE
How to use tesseract collision for collision checking ？

### DISCUSSION THREAD

**Discussion by lanmo0923** [tesseract-robotics/tesseract#890] [Q&A] [answered]
I want to use tesseract collision for collision checking, but I can't find the specific documentation, only in finding some information in https://tesseract-docs.readthedocs.io/en/latest/
 is there any documentation I didn't notice? Can you give me some specific practical guidance?
Thank you all!

**Comment by Levi-Armstrong** [comment, marked as answer]
If you are wanting to use it directly, then tesseract_collision has an example showing basic usage found [here](https://github.com/tesseract-robotics/tesseract/blob/master/tesseract_collision/examples/box_box_example.cpp). If you wanting to use it with loading a urdf and srdf then  I recommend looking at the tesseract_environment examples.

**Comment by lanmo0923** [comment]
Thank you, I want to import the urdf of the robot, specify the collision pair and use it, and visualize it, similar to the figure below, how to visualize the returned collision results？
![image](https://github.com/tesseract-robotics/tesseract/assets/49888704/b511bc7e-863e-4d08-8886-7742c2ae52d1)

**Reply by lanmo0923** [reply]
hello,I try to implement environment collisions in python  refer to the  [example](https://github.com/tesseract-robotics/tesseract/blob/master/tesseract_environment/test/tesseract_environment_collision.cpp).But collision results are always empty.This problem has been bothering me for a few days, can you help me find the code error?  @Levi-Armstrong 


```
class TesseractSupportResourceLocator(ResourceLocator):
    def __init__(self):
        super().__init__()

    def locateResource(self, url):
        try:
            try:
                if os.path.exists(url):
                    return SimpleLocatedResource(url, url, self)
            except:
                pass
            url_match = re.match(r"^package:\/\/tesseract_support\/(.*)$", url)
            if (url_match is None):
                print("url_match failed")
                return None
            if not "TESSERACT_SUPPORT_DIR" in os.environ:
                return None
            tesseract_support = os.environ["TESSERACT_SUPPORT_DIR"]
            filename = os.path.join(tesseract_support, os.path.normpath(url_match.group(1)))
            ret = SimpleLocatedResource(url, filename, self)
            return ret
        except:
            traceback.print_exc()
abb_irb2400_urdf_fname = FilesystemPath(os.path.join(TESSERACT_SUPPORT_DIR, "urdf", "abb_irb2400.urdf"))
abb_irb2400_srdf_fname = FilesystemPath(os.path.join(TESSERACT_SUPPORT_DIR, "urdf", "abb_irb2400.srdf"))
t_env = Environment()
locator = TesseractSupportResourceLocator()
t_env.init(abb_irb2400_urdf_fname, abb_irb2400_srdf_fname, locator)
manip_info = ManipulatorInfo()
manip_info.tcp_frame = "tool0"
manip_info.manipulator = "manipulator"
manip_info.working_frame = "base_link"

viewer = TesseractViewer()
viewer.update_environment(t_env, [0, 0, 0])
joint_names = ["joint_%d" % (i + 1) for i in range(6)]
viewer.update_joint_positions(joint_names, np.array([1, -.2, .01, .3, -.5, 1]))


config=CollisionCheckConfig()
config.longest_valid_segment_length=0.1
config.contact_request.type=tesseract_collision.ContactTestType_CLOSEST
config.type=tesseract_collision.CollisionEvaluatorType_DISCRETE
margin_data = CollisionMarginData(0)
config.contact_manager_config.margin_data = margin_data


checker=t_env.getDiscreteContactManager()
collision_data = ContactResultMap()
checker.contactTest(collision_data,config.contact_request)
print(collision_data.empty())
```

**Reply by Levi-Armstrong** [reply]
So I tested the joint position using Tesseract Studio and nothing is in collision using the joint state and contact threshold.
![image](https://github.com/tesseract-robotics/tesseract/assets/9803128/3e1ea7bd-7f05-4a24-81e1-addc9db1a340)

**Reply by Levi-Armstrong** [reply]
The issue is that you are using a contact margin of 0. Change the contact margin to 1 and you should get the following contacts.
![image](https://github.com/tesseract-robotics/tesseract/assets/9803128/51917240-6ac0-44bf-ba45-dbe046a0ccc7)

**Reply by Levi-Armstrong** [reply]
Just created a MR fixing the arrow visualization.
![image](https://github.com/tesseract-robotics/tesseract/assets/9803128/88a75b98-829c-4a38-8827-ad7a721948f1)

**Reply by lanmo0923** [reply]
@Levi-Armstrong Thank you very much.I set the contact margin to 1 and obtained the results.
`checker.setCollisionMarginData(tesseract_common.CollisionMarginData(1))`
After your help, I feel that the tesseract_qt is effective in helping me to identify problems with the program. But  my development environment does not work with ros, can I use it without ros?

**Reply by Levi-Armstrong** [reply]
Yes, tesseract at its core is designed to be ROS agnostic. The tesseract_qt and Tesseract Studio is fairly new so not sure how you need to use it but it should work. If you can explain how you would like to use I can let you know if it will work right away. One example I could see is some may want the functionality to launch Tesseract Studio from there application providing there environment object, but not sure if this is what you are looking for.

**Reply by Levi-Armstrong** [reply]
Also, Tesseract Studio leverages plugins so you could write your own plugin if you are suing another middle ware to provide similar functionality that tesseract_qt_ros package provides which is part of an open PR on tesseract_ros.

**Reply by lanmo0923** [reply]
@Levi-Armstrong  Thank you for your reply, I am a beginner of tesseract and I am currently trying to use tesseract in python to do collision detection and motion planning for manipulators. I would like to use Tesseract Studio to import the environment to get quick and reliable results to help me verify the correctness of my code or find problems in my code.I read the description of tesseract_qt carefully but didn't find how to install it, can you give me some advice on how to build and use these tools?

**Reply by Levi-Armstrong** [reply]
You will need to clone these three repositories tesseract, tesseract_planning and tesseract_qt into your workspace. Then you should be able to run catkin build.

**Reply by lanmo0923** [reply]
@Levi-Armstrong  Thank you for your reply,In my application, only binary collision results are needed, and when I  query the document, I found that obtaining binary collision results is much faster than obtaining distance information.
![image](https://github.com/tesseract-robotics/tesseract/assets/49888704/5aa0de92-22f4-45d4-b75b-9957a084a762)

 But I have browsed through the documentation, examples, and discussions and have not found any relevant APIs. How can I implement it？

**Reply by Levi-Armstrong** [reply]
In your contact request, set type to FIRST, calculate_penetration to false, calculate_distance to false.

**Reply by lanmo0923** [reply]
I tried to use Tesseract Studio, but after loading ManipulationWidget, it is not working and doesn't show the slider, is there something wrong with the way I loaded it? In fact, I don't understand how to set component info when loading plugins. This problem has been bothering me for some time and I can't find a solution. Can you give me some advice?   @Levi-Armstrong 

![image](https://github.com/tesseract-robotics/tesseract/assets/49888704/b82622c6-a9ff-4b7e-b81f-47c54b2e0be6)

![image](https://github.com/tesseract-robotics/tesseract/assets/49888704/f38d9012-da8b-46f1-a06f-3889e6455fe1)

**Comment by Levi-Armstrong** [comment]
I recommend loading the config found in tesseract_qt/studio/config to start and review the file. The component infos are used to associated separate plugins to one another. The plugins communication using Qt Event Filters and this is used to know if a plugin should operate on an event. 

In your case you would want to probably use the same component info for both the environment and manipulation plugin so they are associated with each other. Then way when you move the slider the data in the environment widget get updated. Also right now they should all have the same scene name. In the future it will be possible to have multiple which would result in a different render table. A case where you might have multiple is if you wanted to load multiple environment widgets and load different urdf or if using the ros environment plugin associated with a different topic.

**Reply by Levi-Armstrong** [reply]
Also on the manipulation widget do not forget to click the reload button. The reload is to pull an updates from the associated environment where say links were moved or changes.

**Reply by lanmo0923** [reply]
@Levi-Armstrong Thanks for your detailed explanation. I tried the solution, but when I clicked reload button, some errors appeared: libEGL warning: 

`Not allowed to force software rendering when API explicitly selects a hardware device.`

I found that since I am using wsl, there are some problems with using hardware rendering with [gzsim.](https://github.com/gazebosim/gz-sim/issues/920) I was wondering if it is possible to unforce hardware rendering in the code or use software rendering? But I can't find a modification when I look at the source code. Can you tell me where to change it if it is possible?

**Reply by Levi-Armstrong** [reply]
Did you try exporting this `export LIBGL_ALWAYS_SOFTWARE=true` before running? It looks like another option is to tell it to use ogre versus ogre2.

**Reply by lanmo0923** [reply]
Yes, I tried both software rendering and hardware rendering and got different errors respectively.
`Not allowed to force software rendering when API explicitly selects a hardware device.`
`libEGL warning: MESA-LOADER: failed to open vgem:`

**Comment by Levi-Armstrong** [comment]
Not sure. I recommend posting your issue on that thread and see what they say.

**Reply by lanmo0923** [reply]
Ok, thanks for your reply
